jubatus/core/bandit/epsilon_decreasing.cpp

```diff
@@ -0,0 +1,126 @@
+// modify it under the terms of the GNU Lesser General Public
+// License version 2.1 as published by the Free Software Foundation.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+#include "epsilon_decreasing.hpp"
+
+#include <string>
+#include <vector>
+#include <limits>
+#include <algorithm>
+
+#include "../common/exception.hpp"
+#include "../framework/packer.hpp"
+#include "../common/version.hpp"
+
+namespace jubatus {
+namespace core {
+namespace bandit {
+
+epsilon_decreasing::epsilon_decreasing(const config& conf)
+    : d_(conf.d), s_(conf.assume_unrewarded) {
+  if (conf.d <= 0 || 1 < conf.d) {
+    throw JUBATUS_EXCEPTION(
+        common::invalid_parameter("0 < d <= 1"));
+  }
+  if (conf.seed) {
+    if (*conf.seed < 0 || std::numeric_limits<uint32_t>::max() < *conf.seed) {
+      throw JUBATUS_EXCEPTION(
+          common::config_exception() << common::exception::error_message(
+              "seed must be within unsigned 32 bit integer"));
+    }
+    rand_ = jubatus::util::math::random::mtrand(*conf.seed);
+  }
+}
+
+std::string epsilon_decreasing::select_arm(const std::string& player_id) {
+  const std::vector<std::string>& arms = s_.get_arm_ids();
+  if (arms.empty()) {
+    throw JUBATUS_EXCEPTION(
+        common::exception::runtime_error("arm is not registered"));
+  }
+  std::string result;
+  int total_trial = s_.get_total_trial_count(player_id) + 1;
+  double log_total_trial = std::log(total_trial);
+  double eps = 5 * arms.size() * (log_total_trial / total_trial) / (d_ * d_);
+  double dec_eps = std::min(1.0, eps);
+  if (rand_.next_double() < dec_eps) {
+    // exploration
+    result = arms[rand_.next_int(arms.size())];
+  } else {
+    // exploitation
+    result = arms[0];
+    double exp_max = s_.get_expectation(player_id, arms[0]);
+    for (size_t i = 1; i < arms.size(); ++i) {
+      double exp = s_.get_expectation(player_id, arms[i]);
+      if (exp > exp_max) {
+        result = arms[i];
+        exp_max = exp;
+      }
+    }
+  }
+  s_.notify_selected(player_id, result);
+  return result;
+}
+
+bool epsilon_decreasing::register_arm(const std::string& arm_id) {
+  return s_.register_arm(arm_id);
+}
+bool epsilon_decreasing::delete_arm(const std::string& arm_id) {
+  return s_.delete_arm(arm_id);
+}
+
+bool epsilon_decreasing::register_reward(const std::string& player_id,
+                                     const std::string& arm_id,
+                                     double reward) {
+  return s_.register_reward(player_id, arm_id, reward);
```

Comment:
In the original paper,  the reward is converted according to the probability of selecting the arm.
Is there any reason you have removed the conversion?
I'm not quite sure the effect of removing it.